src/gt4py/next/new_type_system/traits.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import annotations
+
+import abc
+from typing import Any, Optional
+from collections.abc import Sequence
+
+import dataclasses
+
+
+class Trait:
+    ...
+
+
+class SignednessTrait(Trait):
+    """Specifies the signedness of a type."""
+    @abc.abstractmethod
+    def is_signed(self) -> bool:
+        ...
+
+
+class FromTrait(Trait):
+    """
+    Specifies if a type is constructible FROM another type.
+
+    This trait represents explicit conversions and an implementation should return
+    True if the conversion is possible, even if it is losing arithmetic precision.
+    """
+
+    @abc.abstractmethod
+    def is_constructible_from(self, ty: Any) -> bool:
```

Comment:
Yeah, this is a major problem. How can I use `types.Type` in `traits.py` for these annotations so that I don't create cycle? `types.py` must import `traits.py`, and not only for annotations.

Reply:
Solved by moving shared object into a separate file.